File: truss-compiler/src/main/java/truss/compiler/MessageCollectionScope.java
package truss.compiler;

import org.apache.commons.lang.Validate;

public class MessageCollectionScope {
    private MessageCollectionScope() {
    }

    private static final ThreadLocal<MessageCollection> CURRENT = new ThreadLocal<>();

    public static MessageCollection getCurrent() {
        return CURRENT.get();
    }

    public static void addMessage(Message message) {
        Validate.notNull(message, "message");

        MessageCollection messages = getCurrent();

        assert messages != null;

        messages.add(message);
    }

    public static Closeable create(MessageCollection messages) {
        Validate.notNull(messages, "messages");

        CURRENT.set(messages);

        return new Closeable() {
            @Override
            public void close() {
                CURRENT.remove();
            }
        };
    }
}

File: Truss.Compiler/MessageCollectionScope.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace Truss.Compiler {
    public static class MessageCollectionScope {
        [ThreadStatic]
        private static MessageCollection _current;

        public static MessageCollection Current {
            get { return _current; }
        }

        public static void AddMessage(Message message) {
            if (message == null) {
                throw new ArgumentNullException("message");
            }

            var messages = Current;

            Debug.Assert(messages != null);

            messages.Add(message);
        }

        public static IDisposable Create(MessageCollection messages) {
            if (messages == null) {
                throw new ArgumentNullException("messages");
            }

            _current = messages;

            return new Handle();
        }

        private class Handle : IDisposable {
            private bool _disposed;

            public void Dispose() {
                if (!_disposed) {
                    _current = null;

                    _disposed = true;
                }
            }
        }
    }
}
